src/util/taproot.rs

```diff
@@ -451,19 +452,27 @@ impl TaprootBuilder {
 
     /// Creates a [`TaprootSpendInfo`] with the given internal key.
     ///
-    // TODO: in a future breaking API change, this no longer needs to return result
+    /// Returns the unmodified builder as Err if the builder is not finalized.
+    /// See also [`TaprootBuilder::is_finalized`]
     pub fn finalize<C: secp256k1::Verification>(
         mut self,
         secp: &Secp256k1<C>,
         internal_key: UntweakedPublicKey,
-    ) -> Result<TaprootSpendInfo, TaprootBuilderError> {
-        match self.branch.pop() {
-            None => Ok(TaprootSpendInfo::new_key_spend(secp, internal_key, None)),
-            Some(Some(node)) => {
-                Ok(TaprootSpendInfo::from_node_info(secp, internal_key, node))
+    ) -> Result<TaprootSpendInfo, TaprootBuilder> {
+        match self.branch.len() {
+            0 => Ok(TaprootSpendInfo::new_key_spend(secp, internal_key, None)),
+            1 => {
+                if let Some(Some(node)) = self.branch.pop() {
+                    Ok(TaprootSpendInfo::from_node_info(secp, internal_key, node))
+                } else {
+                    // This should be unreachable as
+                    // - The builder invariant guarantees that the last element is always Some
+                    // - The size of vector is checked to be 1
+                    // But still we return an error in order to avoid a panic reasoning.
+                    Err(self)
```

Comment:
In 435eb9ba34cda0650bf2aac0192b9680c0f82a30:

Honestly I'd prefer we just `.expect` here. We checked the length of the vector literally one line above, we know `pop` will succeed.